// @mastra/server/dist/chunk-UCTEMO2Q.cjs
'use strict';

var chunkYWLUOY3D_cjs = require('./chunk-YWLUOY3D.cjs');
var chunk64U3UDTH_cjs = require('./chunk-64U3UDTH.cjs');
var chunkOCWPVYNI_cjs = require('./chunk-OCWPVYNI.cjs');
var chunk75ZPJI57_cjs = require('./chunk-75ZPJI57.cjs');
var web = require('stream/web');
var di = require('@mastra/core/di');

// src/server/handlers/vNextWorkflows.ts
var vNextWorkflows_exports = {};
chunk75ZPJI57_cjs.__export(vNextWorkflows_exports, {
  createVNextWorkflowRunHandler: () => createVNextWorkflowRunHandler,
  getVNextWorkflowByIdHandler: () => getVNextWorkflowByIdHandler,
  getVNextWorkflowRunByIdHandler: () => getVNextWorkflowRunByIdHandler,
  getVNextWorkflowRunsHandler: () => getVNextWorkflowRunsHandler,
  getVNextWorkflowsHandler: () => getVNextWorkflowsHandler,
  resumeAsyncVNextWorkflowHandler: () => resumeAsyncVNextWorkflowHandler,
  resumeVNextWorkflowHandler: () => resumeVNextWorkflowHandler,
  startAsyncVNextWorkflowHandler: () => startAsyncVNextWorkflowHandler,
  startVNextWorkflowRunHandler: () => startVNextWorkflowRunHandler,
  watchVNextWorkflowHandler: () => watchVNextWorkflowHandler
});
async function getVNextWorkflowsHandler({ mastra }) {
  try {
    const workflows = mastra.vnext_getWorkflows({ serialized: false });
    const _workflows = Object.entries(workflows).reduce((acc, [key, workflow]) => {
      acc[key] = {
        name: workflow.name,
        steps: Object.entries(workflow.steps).reduce((acc2, [key2, step]) => {
          acc2[key2] = {
            id: step.id,
            description: step.description,
            inputSchema: step.inputSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(step.inputSchema)) : void 0,
            outputSchema: step.outputSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(step.outputSchema)) : void 0,
            resumeSchema: step.resumeSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(step.resumeSchema)) : void 0,
            suspendSchema: step.suspendSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(step.suspendSchema)) : void 0
          };
          return acc2;
        }, {}),
        stepGraph: workflow.serializedStepGraph,
        inputSchema: workflow.inputSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(workflow.inputSchema)) : void 0,
        outputSchema: workflow.outputSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(workflow.outputSchema)) : void 0
      };
      return acc;
    }, {});
    return _workflows;
  } catch (error) {
    throw new chunkOCWPVYNI_cjs.HTTPException(500, { message: error?.message || "Error getting workflows" });
  }
}
async function getVNextWorkflowByIdHandler({ mastra, workflowId }) {
  try {
    if (!workflowId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "Workflow ID is required" });
    }
    const workflow = mastra.vnext_getWorkflow(workflowId);
    if (!workflow) {
      throw new chunkOCWPVYNI_cjs.HTTPException(404, { message: "Workflow not found" });
    }
    return {
      steps: Object.entries(workflow.steps).reduce((acc, [key, step]) => {
        acc[key] = {
          id: step.id,
          description: step.description,
          inputSchema: step.inputSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(step.inputSchema)) : void 0,
          outputSchema: step.outputSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(step.outputSchema)) : void 0,
          resumeSchema: step.resumeSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(step.resumeSchema)) : void 0,
          suspendSchema: step.suspendSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(step.suspendSchema)) : void 0
        };
        return acc;
      }, {}),
      name: workflow.name,
      stepGraph: workflow.serializedStepGraph,
      inputSchema: workflow.inputSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(workflow.inputSchema)) : void 0,
      outputSchema: workflow.outputSchema ? chunkYWLUOY3D_cjs.stringify(chunkYWLUOY3D_cjs.esm_default(workflow.outputSchema)) : void 0
    };
  } catch (error) {
    throw new chunkOCWPVYNI_cjs.HTTPException(500, { message: error?.message || "Error getting workflow" });
  }
}
async function getVNextWorkflowRunByIdHandler({
  mastra,
  workflowId,
  runId
}) {
  try {
    if (!workflowId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "Workflow ID is required" });
    }
    if (!runId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "Run ID is required" });
    }
    const workflow = mastra.vnext_getWorkflow(workflowId);
    if (!workflow) {
      throw new chunkOCWPVYNI_cjs.HTTPException(404, { message: "Workflow not found" });
    }
    const run = await workflow.getWorkflowRunById(runId);
    if (!run) {
      throw new chunkOCWPVYNI_cjs.HTTPException(404, { message: "Workflow run not found" });
    }
    return run;
  } catch (error) {
    throw new chunkOCWPVYNI_cjs.HTTPException(500, { message: error?.message || "Error getting workflow run" });
  }
}
async function createVNextWorkflowRunHandler({
  mastra,
  workflowId,
  runId: prevRunId
}) {
  try {
    if (!workflowId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "Workflow ID is required" });
    }
    const workflow = mastra.vnext_getWorkflow(workflowId);
    if (!workflow) {
      throw new chunkOCWPVYNI_cjs.HTTPException(404, { message: "Workflow not found" });
    }
    const run = workflow.createRun({ runId: prevRunId });
    return { runId: run.runId };
  } catch (error) {
    throw new chunkOCWPVYNI_cjs.HTTPException(500, { message: error?.message || "Error creating workflow run" });
  }
}
async function startAsyncVNextWorkflowHandler({
  mastra,
  runtimeContext,
  workflowId,
  runId,
  inputData,
  runtimeContextFromRequest
}) {
  try {
    if (!workflowId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "Workflow ID is required" });
    }
    const workflow = mastra.vnext_getWorkflow(workflowId);
    if (!workflow) {
      throw new chunkOCWPVYNI_cjs.HTTPException(404, { message: "Workflow not found" });
    }
    const finalRuntimeContext = new di.RuntimeContext([
      ...Array.from(runtimeContext?.entries() ?? []),
      ...Array.from(Object.entries(runtimeContextFromRequest ?? {}))
    ]);
    const _run = workflow.createRun({ runId });
    const result = await _run.start({
      inputData,
      runtimeContext: finalRuntimeContext
    });
    return result;
  } catch (error) {
    throw new chunkOCWPVYNI_cjs.HTTPException(500, { message: error?.message || "Error executing workflow" });
  }
}
async function startVNextWorkflowRunHandler({
  mastra,
  runtimeContext,
  workflowId,
  runId,
  inputData,
  runtimeContextFromRequest
}) {
  try {
    if (!workflowId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "Workflow ID is required" });
    }
    if (!runId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "runId required to start run" });
    }
    const workflow = mastra.vnext_getWorkflow(workflowId);
    const run = await workflow.getWorkflowRunById(runId);
    if (!run) {
      throw new chunkOCWPVYNI_cjs.HTTPException(404, { message: "Workflow run not found" });
    }
    const finalRuntimeContext = new di.RuntimeContext([
      ...Array.from(runtimeContext?.entries() ?? []),
      ...Array.from(Object.entries(runtimeContextFromRequest ?? {}))
    ]);
    const _run = workflow.createRun({ runId });
    void _run.start({
      inputData,
      runtimeContext: finalRuntimeContext
    });
    return { message: "Workflow run started" };
  } catch (e) {
    return chunk64U3UDTH_cjs.handleError(e, "Error starting workflow run");
  }
}
async function watchVNextWorkflowHandler({
  mastra,
  workflowId,
  runId
}) {
  try {
    if (!workflowId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "Workflow ID is required" });
    }
    if (!runId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "runId required to watch workflow" });
    }
    const workflow = mastra.vnext_getWorkflow(workflowId);
    const run = await workflow.getWorkflowRunById(runId);
    if (!run) {
      throw new chunkOCWPVYNI_cjs.HTTPException(404, { message: "Workflow run not found" });
    }
    const _run = workflow.createRun({ runId });
    let unwatch;
    let asyncRef = null;
    const stream = new web.ReadableStream({
      start(controller) {
        unwatch = _run.watch(({ type, payload, eventTimestamp }) => {
          controller.enqueue(JSON.stringify({ type, payload, eventTimestamp, runId }));
          if (asyncRef) {
            clearImmediate(asyncRef);
            asyncRef = null;
          }
          asyncRef = setImmediate(async () => {
            const runDone = payload.workflowState.status !== "running";
            if (runDone) {
              controller.close();
              unwatch?.();
            }
          });
        });
      },
      cancel() {
        unwatch?.();
      }
    });
    return stream;
  } catch (error) {
    return chunk64U3UDTH_cjs.handleError(error, "Error watching workflow");
  }
}
async function resumeAsyncVNextWorkflowHandler({
  mastra,
  workflowId,
  runId,
  body,
  runtimeContext,
  runtimeContextFromRequest
}) {
  try {
    if (!workflowId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "Workflow ID is required" });
    }
    if (!runId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "runId required to resume workflow" });
    }
    if (!body.step) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "step required to resume workflow" });
    }
    const workflow = mastra.vnext_getWorkflow(workflowId);
    const run = await workflow.getWorkflowRunById(runId);
    if (!run) {
      throw new chunkOCWPVYNI_cjs.HTTPException(404, { message: "Workflow run not found" });
    }
    const finalRuntimeContext = new di.RuntimeContext([
      ...Array.from(runtimeContext?.entries() ?? []),
      ...Array.from(Object.entries(runtimeContextFromRequest ?? {}))
    ]);
    const _run = workflow.createRun({ runId });
    const result = await _run.resume({
      step: body.step,
      resumeData: body.resumeData,
      runtimeContext: finalRuntimeContext
    });
    return result;
  } catch (error) {
    return chunk64U3UDTH_cjs.handleError(error, "Error resuming workflow step");
  }
}
async function resumeVNextWorkflowHandler({
  mastra,
  workflowId,
  runId,
  body,
  runtimeContext
}) {
  try {
    if (!workflowId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "Workflow ID is required" });
    }
    if (!runId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "runId required to resume workflow" });
    }
    if (!body.step) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "step required to resume workflow" });
    }
    const workflow = mastra.vnext_getWorkflow(workflowId);
    const run = await workflow.getWorkflowRunById(runId);
    if (!run) {
      throw new chunkOCWPVYNI_cjs.HTTPException(404, { message: "Workflow run not found" });
    }
    const _run = workflow.createRun({ runId });
    void _run.resume({
      step: body.step,
      resumeData: body.resumeData,
      runtimeContext
    });
    return { message: "Workflow run resumed" };
  } catch (error) {
    return chunk64U3UDTH_cjs.handleError(error, "Error resuming workflow");
  }
}
async function getVNextWorkflowRunsHandler({
  mastra,
  workflowId,
  fromDate,
  toDate,
  limit,
  offset,
  resourceId
}) {
  try {
    if (!workflowId) {
      throw new chunkOCWPVYNI_cjs.HTTPException(400, { message: "Workflow ID is required" });
    }
    const workflow = mastra.vnext_getWorkflow(workflowId);
    const workflowRuns = await workflow.getWorkflowRuns({ fromDate, toDate, limit, offset, resourceId }) || {
      runs: [],
      total: 0
    };
    return workflowRuns;
  } catch (error) {
    return chunk64U3UDTH_cjs.handleError(error, "Error getting workflow runs");
  }
}

exports.createVNextWorkflowRunHandler = createVNextWorkflowRunHandler;
exports.getVNextWorkflowByIdHandler = getVNextWorkflowByIdHandler;
exports.getVNextWorkflowRunByIdHandler = getVNextWorkflowRunByIdHandler;
exports.getVNextWorkflowRunsHandler = getVNextWorkflowRunsHandler;
exports.getVNextWorkflowsHandler = getVNextWorkflowsHandler;
exports.resumeAsyncVNextWorkflowHandler = resumeAsyncVNextWorkflowHandler;
exports.resumeVNextWorkflowHandler = resumeVNextWorkflowHandler;
exports.startAsyncVNextWorkflowHandler = startAsyncVNextWorkflowHandler;
exports.startVNextWorkflowRunHandler = startVNextWorkflowRunHandler;
exports.vNextWorkflows_exports = vNextWorkflows_exports;
exports.watchVNextWorkflowHandler = watchVNextWorkflowHandler;
